app/main.py

```diff
@@ -1,214 +1,71 @@
-
-    # apply armour
-    lancelot["protection"] = 0
-    for a in lancelot["armour"]:
-        lancelot["protection"] += a["protection"]
-
-    # apply weapon
-    lancelot["power"] += lancelot["weapon"]["power"]
-
-    # apply potion if exist
-    if lancelot["potion"] is not None:
-        if "power" in lancelot["potion"]["effect"]:
-            lancelot["power"] += lancelot["potion"]["effect"]["power"]
-
-        if "protection" in lancelot["potion"]["effect"]:
-            lancelot["protection"] += lancelot["potion"]["effect"]["protection"]
-
-        if "hp" in lancelot["potion"]["effect"]:
-            lancelot["hp"] += lancelot["potion"]["effect"]["hp"]
-
-    # arthur
-    arthur = knightsConfig["arthur"]
-
-    # apply armour
-    arthur["protection"] = 0
-    for a in arthur["armour"]:
-        arthur["protection"] += a["protection"]
-
-    # apply weapon
-    arthur["power"] += arthur["weapon"]["power"]
-
-    # apply potion if exist
-    if arthur["potion"] is not None:
-        if "power" in arthur["potion"]["effect"]:
-            arthur["power"] += arthur["potion"]["effect"]["power"]
-
-        if "protection" in arthur["potion"]["effect"]:
-            arthur["protection"] += arthur["potion"]["effect"]["protection"]
-
-        if "hp" in arthur["potion"]["effect"]:
-            arthur["hp"] += arthur["potion"]["effect"]["hp"]
-
-    # mordred
-    mordred = knightsConfig["mordred"]
-
-    # apply armour
-    mordred["protection"] = 0
-    for a in mordred["armour"]:
-        mordred["protection"] += a["protection"]
-
-    # apply weapon
-    mordred["power"] += mordred["weapon"]["power"]
+from typing import Dict, Any
+
+
+class Knight:
+    def __init__(
+            self,
+            name: str,
+            power: int,
+            hp: int,
+            armour: list[Dict[str, int]],
+            weapon: dict, potion:
+            dict = None
+    ) -> None:
+        self.name = name
+        self.power = power
+        self.hp = hp
+        self.armour = armour
+        self.weapon = weapon
+        self.potion = potion if potion else {}
+
+    def get_protection(self) -> int:
+        return sum(part["protection"] for part in self.armour)
+
+    def get_effective_power(self) -> int:
+        weapon_power = self.weapon["power"]
+        potion_power = self.potion.get("effect", {}).get("power", 0)
+        return self.power + weapon_power + potion_power
+
+    def get_effective_hp(self) -> int:
+        potion_hp = self.potion.get("effect", {}).get("hp", 0)
+        return self.hp + potion_hp
+
+    def get_effective_protection(self) -> int:
+        potion_protection = self.potion.get("effect", {}).get("protection", 0)
+        return self.get_protection() + potion_protection
+
+
+def battle(knights_config: Dict[str, Dict[str, Any]]) -> Dict[str, int]:
+    knights = {name: Knight(**data) for name, data in knights_config.items()}
+
+    lancelot = knights["lancelot"]
+    mordred = knights["mordred"]
```

Comment:
The code assumes that 'lancelot' and 'mordred' are always present in `knights_config`. Consider adding error handling to manage cases where these keys might be missing.